src/rules/maxNestingDepthRule.ts

```diff
@@ -0,0 +1,183 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { isFunctionWithBody, isIdentifier } from "tsutils";
+import * as ts from "typescript";
+import * as Lint from "../index";
+
+/**
+ * Rule max-nesting-depth
+ */
+export class Rule extends Lint.Rules.AbstractRule {
+    public static MINIMUM_THRESHOLD = 1;
+    public static DEFAULT_THRESHOLD = 4;
+    public static THRESHOLD_EXAMPLE = 7;
+
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "max-nesting-depth",
+        description: "Enforces a maximum threshold of nesting depth.",
+        descriptionDetails: Lint.Utils.dedent`
+            The nesting depth metric is assessed for each function of any type. It is computed by
+            determining the nesting levels of the control statements within the function. A starting
+            value of 0 is assigned, this value is incremented each time a new control statement is
+            entered, then the value is decremented when the control statement is left. The following
+            control statements contribute to the nesting depth computation:
+            * \`if\` statements
+            * \`?:\` ternary operators
+            * \`switch\` statements
+            * \`for\`, \`for in\` and \`for of\` loops
+            * \`while\` and \`do while\` loops
+            * \`try\` \`catch\` and \`finally\` blocks
+
+            The [\`max-nesting-depth\`](../max-nesting-depth) metric and the
+            [\`cyclomatic-complexity\`](../cyclomatic-complexity) metric are both computed using the
+            control statements. The cyclomatic complexity metric simply counts the control statements
+            of the function until its threshold is reached whereas the nesting depth metric increase
+            and decrease each time a control statement is entered and left in the function until its
+            threshold is reached.`,
+        rationale: Lint.Utils.dedent`
+            Maximum nesting depth is a code metric which indicates how deep control statements are
+            nested in a function. This metric quickly reveals the functions whose code is made
+            difficult to read by an excessive nesting of the control statements.
+
+            It's better to have smaller, single-purpose functions with self-documenting names.`,
+        optionsDescription: Lint.Utils.dedent`
+            An optional upper limit for maximum nesting depth can be specified. If no limit option
+            is provided a default value of ${Rule.DEFAULT_THRESHOLD} will be used.`,
+        options: {
```

Comment:
TSLint's been moving towards having options as objects with members, rather than single primitives. They're a lot easier to add more options to later on and a bit more clear in `tslint.json`s. You'll want to switch this to `type: "object"`; see `memberOrderingRule.ts` and `noShawdowedVariableRule.ts` as examples.